realm/realm-library/src/androidTest/java/io/realm/RealmCacheTest.java

```diff
@@ -0,0 +1,291 @@
+        // If the configuration cache wasn't cleared this would fail as we would detect two
+        // configurations with 2 different passwords pointing to the same file.
+        RealmConfiguration newConfig = new RealmConfiguration.Builder(getContext())
+                .name(REALM_NAME)
+                .encryptionKey(newPassword)
+                .migration(realmMigration)
+                .schema(StringOnly.class)
+                .build();
+
+        testRealm = Realm.getInstance(newConfig);
+        assertNotNull(testRealm);
+        testRealm.close();
+    }
+
+    // Tests that if the same Realm file is opened on multiple threads, we only need to validate the schema on the first thread.
+    public void testValidateSchemasOverThreads() throws InterruptedException, TimeoutException, ExecutionException {
+        final RealmConfiguration config = TestHelper.createConfiguration(getContext(), "foo");
+        Realm.deleteRealm(config);
+
+        final CountDownLatch bgThreadLocked = new CountDownLatch(1);
+        final CountDownLatch mainThreadDone = new CountDownLatch(1);
+
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realm = Realm.getInstance(config);
+                realm.beginTransaction();
+                bgThreadLocked.countDown();
+                try {
+                    mainThreadDone.await(5, TimeUnit.SECONDS);
+                } catch (InterruptedException ignored) {
+                }
+                realm.close();
+            }
+        }).start();
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        Future<Boolean> future = executorService.submit(new Callable<Boolean>() {
+            @Override
+            public Boolean call() throws Exception {
+                Realm realm = Realm.getInstance(config);
+                realm.close();
+                mainThreadDone.countDown();
+                return true;
+            }
+        });
+
+        bgThreadLocked.await(2, TimeUnit.SECONDS);
+        assertTrue(future.get(10, TimeUnit.SECONDS));
+    }
+
+    public void testDifferentThreadsDifferentInstance() throws InterruptedException {
+        final CountDownLatch openLatch = new CountDownLatch(1);
+        final CountDownLatch closeLatch = new CountDownLatch(1);
+
+        final Realm realmA = Realm.getInstance(defaultConfig);
+
+        Thread thread = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                Realm realmB = Realm.getInstance(defaultConfig);
+                assertFalse(realmA == realmB);
+                RealmCache.invokeWithGlobalRefCount(defaultConfig,
+                        new TestHelper.ExpectedCountCallback(RealmCacheTest.this, 2));
+                openLatch.countDown();
+                realmB.close();
+                closeLatch.countDown();
+            }
+        });
+        thread.start();
+
+        openLatch.await();
```

Comment:
what's the purpose of `openLatch`? you can just use `closeLatch` to make the `InstrumentationThread` block,
you can even use `thread.join()` without any `CountDownLatch`